### DISCUSSION TITLE
Potential fix for race condition on shutdown

### DISCUSSION THREAD

**Discussion by WeaselKing** [buttplugio/buttplug-unity#11] [Ideas]
I was having a race condition on shutdown with the server process. This was making an exception happen on shutdown when it tried to kill the server process which had already exited, but it only happened about half the time I closed the program.

I stopped getting the exception by changing StopServer in ButtplugUnityHelper.cs to:
```
public static void StopServer(bool exited = false)
    {
      if (ButtplugUnityHelper.serverProcess == null)
      {
        return;
      }
      // Remove the exit event handler before killing the process, otherwise the
      // event handler will call StopServer and we'll end up in an infinite
      // event loop (heh).
      ButtplugUnityHelper.serverProcess.Exited -= ButtplugUnityHelper.OnServerExit;
      ButtplugUnityHelper.serverProcess.OutputDataReceived -= ButtplugUnityHelper.OnServerLogMessage;
      if (!exited)
      {
                if (!serverProcess.HasExited)
                {
                    ButtplugUnityHelper.serverProcess.Kill();
                }
      }
      ButtplugUnityHelper.serverProcess = null;
    }
```

I assume it has something to do with Unity shenanigans, but making that IF statement 
```
if (!exited && !serverProcess.HasExited)
```
Causes an exception "Win32Exception: Operation has completed successfully" which is super strange from my perception.

**Comment by qdot** [comment]
Oh dear god. Yeah I wouldn't put it past there being some sort of weirdness there, I've run into *so* much odd shit via buttplug-unity. I'll see about trying to get this integrated soon.
